feat: add basic game simulation

adam-narozniak opened the pull request:
* simple result display based on the game results (no signalization of opponent choice)
* more logic is needed for many games 
*I didn't use the Game and Round (refactoring and/or some interface)

leonardopennino reviewed (changes requested):
#### There are some problems that need to be addressed with this feature: 
- The game fragment should not handle logic. You must create an interface for the game, then a class who implements the interface and finally the game fragment must use that class for game logic.

- Using a different callback for each button press, although it surely works fine is not portable, consider using one single function for every button and multiplex based on the view parameter sent by android api.

adam-narozniak commented:
There are the game and round classes and logic. Though there are 1) currently in change 2) I think there is general refactoring needed of them (I'll talk more about that with Gaetan).
Therefore mostly not to waste time I haven't done a full implementation connected with that part. (as well as the game itself, a single-round game is kind of supported - meaning you get the result displayed without the opponent choice displayed)
I'll make the buttons better result better, it was also a temporal solution.

leonardopennino commented:
It's still missing an abstraction of the game. Abstract the game methods and it will be ok for merging.

leonardopennino reviewed (approved):
The changes are what I had requested, we can merge! As just a little sidenote, I would have considered having just one handler for the game buttons (Rock paper scissor) which call the same function and the function multiplexes the events. But other than that really good work 👍 !

panadax commented:
> I don't like the idea of having logic in the data classes. That's bad for readability and isn't a good practice imo.
> 
> The logic should either take place in FirebaseRepository, or on a dedicated class that manages game state as a whole (named something like GameService.

I think we should have a quick meeting tomorrow to discuss our framework since it seems there are some common questions to all of us to be confused about what is the best practice of code decoupling.

leonardopennino commented:
> I don't like the idea of having logic in the data classes. That's bad for readability and isn't a good practice imo.
> 
> The logic should either take place in FirebaseRepository, or on a dedicated class that manages game state as a whole (named something like GameService.

I agree, I did not mean to put the logic in data classes, I just said it should be decoupled from the fragment. The class of course won't be a data class.

adam-narozniak commented:
Thanks, you're right. Do you think I should use a specific new class, or should it be in a ViewModel?
And sure, let's meet.

adam-narozniak commented:
Hi, I've made changes in two major areas
1) simplified the transitions to the gameResults and cleared the code involved  
2) switched to the newly implemented OfflineGameService based on GameService for the game logic (here I didn't implement many functions from the interface, it's only todo there, maybe you know a better solution, or maybe I should divide the interface into two smaller ones and use only one?)
Let me know what you think, I'll add tests after I know the design is fine

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -0,0 +1,111 @@
+    override val isGameFull: Boolean
+        get() = true
+
+    override suspend fun playHand(hand: Hand) {
+        TODO("Not yet implemented")
+    }
+
+    override fun dispose() {
+        TODO("Not yet implemented")
+    }
+
+    override val isDisposed: Boolean
+        get() = TODO("Not yet implemented")
+
+    override val active: Boolean
+        get() = TODO("Not yet implemented")
+
+    override suspend fun refreshGame(): Game {
+        TODO("Not yet implemented")
+    }
+
+    override val error: FirebaseFirestoreException
+        get() = TODO("Not yet implemented")
+
+    override fun startListening(): GameService {
+        TODO("probably not needed here")
+    }
+
+    override fun stopListening() {
+        TODO("Not yet implemented")
+    }
```
This should be implemented. If there is no need for such a function in your code, just keep it empty.

adam-narozniak replied:
I can do that in case of the functions that don't return but otherwise, I got an error. How do you advice to proceed?

gaetschwartz replied:
Well `playHand` should be implemented. The others are pretty straightforward as well. The game is always active on your end, so just always return true, and for refresh game, the game is always up to do date, so just always return our game.

It's important you carefully read the FirebaseGameService to understand how methods should behave.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -0,0 +1,111 @@
+        get() = _game?.current_round == roundCount
+
+
+    suspend fun determineRoundResult(userHand: Hand): Hand.Result {
+        userCurrentHand = userHand
+        computerCurrentHand = computerPlayer.makeMove()
+        _round = addRound()
+        val scoreList = _round!!.computeScores()
+        var currentUserResult: Hand.Result? = null
+        for (scoreRes in scoreList) {
+            if (scoreRes.uid == uid) {
+                currentUserResult = scoreRes.results[0]
+                break
+            }
+        }
+        return currentUserResult!!
+    }
```
This method shouldn't have side effects, only methods from GameService should drive the game. Notably 
```kotlin
userCurrentHand = userHand
computerCurrentHand = computerPlayer.makeMove()
_round = addRound()
```
 shouldn't be there.

adam-narozniak replied:
Okay, now I'm trying to change it though there are no methods in GameService that would let me do the computation and determine the winner. How do you want me to proceed?

gaetschwartz replied:
It's not up to the GameService to do so. All the data should be in the Game object, thus making the score computable however we need later on.

gaetschwartz commented:
If you're unsure what a method needs to do, check [FirebaseGameService.kt]( https://github.com/SDP-Rock-Paper-Scissors/RockPaperScissors/blob/main/app/src/main/java/ch/epfl/sweng/rps/services/FirebaseGameService.kt#L19-L139)

adam-narozniak commented on app/build.gradle:
```diff
@@ -157,4 +160,8 @@ task jacocoTestReport(type: JacocoReport, dependsOn: ['testDebugUnitTest', 'crea
     finalizedBy jacocoTestReport
+}
+
+dependencies {
+    testImplementation 'org.junit.jupiter:junit-jupiter'
 }
```
Thanks, I guess I accidentally separated it during one of merges

adam-narozniak commented on app/src/androidTest/java/ch/epfl/sweng/rps/GameButtonsTest.kt:
```diff
@@ -14,39 +14,40 @@ import org.hamcrest.Matchers.not
-
-}
\ No newline at end of file
+// temporarily stop the tests, new behaviour needed
+// their state can't be checked, new screen is loaded
+//@RunWith(AndroidJUnit4::class)
+//@SmallTest
+//class GameButtonsTest {
+//    @get:Rule
+//    val testRule = ActivityScenarioRule(MainActivity::class.java)
+//    @Test
+//    fun pressedRock(){
+//        checkPressedButton(R.id.rockRB)
+//    }
+//    @Test
+//    fun pressedPaper(){
+//        checkPressedButton(R.id.paperRB)
+//    }
+//    @Test
+//    fun pressedScissors() {
+//        checkPressedButton(R.id.scissorsRB)
+//    }
+//    private fun checkPressedButton(radioButtonId: Int){
+//        onView(withId(R.id.button_play_one_offline_game)).perform(click())
+//        onView(withId(radioButtonId)).perform(click())
+//        onView(withId(radioButtonId)).check(matches(isChecked()))
+//    }
+//    @Test
+//    fun pressedRockPaperRock(){
+//        onView(withId(R.id.button_play_one_offline_game)).perform(click())
+//        onView(withId(R.id.rockRB)).perform(click())
+//        onView(withId(R.id.paperRB)).perform(click())
+//        onView(withId(R.id.rockRB)).perform(click())
+//        onView(withId(R.id.rockRB)).check(matches(isChecked()))
+//        onView(withId(R.id.paperRB)).check(matches(not(isChecked())))
+//        onView(withId(R.id.scissorsRB)).check(matches(not(isChecked())))
+//    }
+//
+//
+//}
```
"Let me know what you think, I'll add tests after I know the design is fine" - no it was just commented because the behavior of the buttons has completely changed, though I'll probably use some of the logic that exists there.

adam-narozniak commented on app/src/main/java/ch/epfl/sweng/rps/models/ComputerPlayer.kt:
```diff
@@ -0,0 +1,14 @@
+package ch.epfl.sweng.rps.models
+
+abstract class ComputerPlayer (
+    private val _computerPlayerId: String,
+    private val _numberOfPossibleMoves: Int){
+    val computerPlayerId: String
+        get() = _computerPlayerId
+    val numberOfPossibleMoves: Int
+        get() = _numberOfPossibleMoves
+    /**
+     * @return number representing a move from possible range [0, ..., numberOfMoves - 1]
+     */
+    abstract fun makeMove(): Hand
+}
```
sure, I'll change it

adam-narozniak commented on app/src/main/java/ch/epfl/sweng/rps/models/RandomPlayer.kt:
```diff
@@ -0,0 +1,16 @@
+
+import kotlin.random.Random
+
+class RandomPlayer(numberOfPossibleMoves: Int) :
+    ComputerPlayer("randomPlayer", numberOfPossibleMoves) {
+    override fun makeMove(): Hand {
+        val upperRange: Int = numberOfPossibleMoves - 1
+        return when (Random.nextInt(0, upperRange)) {
+            0 -> Hand.ROCK
+            1 -> Hand.PAPER
+            2 -> Hand.SCISSORS
+            else -> throw IllegalStateException("Given number is not correct")
+        }
+    }
+}
```
For different types of games e.g. rock paper scissors lizard spock, where the number of possible moves will be 5

gaetschwartz replied:
That doesn't really make sense to me. If you want to support different hands in the future, I would rather provide the set of allowed hands. 

```kotlin
class RandomPlayer(allowedHands: List<Hand>) :
    ComputerPlayer("randomPlayer", allowedHands) {
    override fun makeMove(): Hand {
        val upperRange: Int = allowedHands.size
        return allowedHands[Random.nextInt(0, upperRange)]
    }
}
```

**Also, not that `Random.nextInt(0, x)` generates a number from 0 to x (EXCLUSIVE). So you shouldn't do `val upperRange: Int = numberOfPossibleMoves - 1`.**

You also don't need to provide 0 as a parameter, as by default `Random.nextInt(x)` generates from 0 until x (EXCLUSIVE).

adam-narozniak replied:
I could do that though the hand is not abstracted, and I need to make the change on the level of ComputerPlaye. I could make an abstract class Choice (which would be empty) and follow the way you described.
Thanks for the nextInt

gaetschwartz replied:
I would say that for now, you should drop the option for having a custom number of moves. We will see how we implement it in the future when we support more. But for now it can be confusing for other people, and doesn't serve any purpose yet.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/models/Game.kt:
```diff
@@ -15,8 +15,8 @@ import com.google.firebase.Timestamp
     val players: List<String>,
-    val rounds: Map<String, Round>,
-    val current_round: Int,
+    val rounds: MutableMap<String, Round>,
+    var current_round: Int,
```
You shouldn't modifiy this class. `MutableMap` extends `Map`, so there is no need to change it to `MutableMap`. For the round, you should use `game.copy()` to have a new instance with your new `current_round`.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/models/Round.kt:
```diff
@@ -4,12 +4,9 @@ import ch.epfl.sweng.rps.models.Hand.Result
+
 data class Round(
-    val hands: Map<String, Hand>,
+    val hands: MutableMap<String, Hand>,
```
Same thing here.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    override val error: FirebaseFirestoreException
+        get() = TODO("Not yet implemented")
```
This should be implemented.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/FirebaseGameService.kt:
```diff
@@ -79,7 +79,7 @@ class FirebaseGameService(
         }
         val round = Round(
-            hands = emptyMap(),
+            hands = mutableMapOf(),
```
Same thing.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -0,0 +1,105 @@
+import com.google.firebase.firestore.FirebaseFirestoreException
+import java.util.*
+
+class OfflineGameService(
+    private val gameId_: String,
+    private val uid: String,
+    private val computerPlayers: List<ComputerPlayer>,
+    private var roundCount: Int
+) : GameService {
```
I don't think these should be the arguments of the class. I would say only the game mode makes sense as the argument. From this you can derive everything you need, generate a game id, etc.

This logic would be most likely be done in `startListening()`.

adam-narozniak replied:
Am I supposed to get the computer players there and userId?

adam-narozniak commented on app/src/main/java/ch/epfl/sweng/rps/models/Game.kt:
```diff
@@ -15,8 +15,8 @@ import com.google.firebase.Timestamp
     val players: List<String>,
-    val rounds: Map<String, Round>,
-    val current_round: Int,
+    val rounds: MutableMap<String, Round>,
+    var current_round: Int,
```
Actually, it doesn't work in a straightforward way on a normal map. The map doesn't have the "put" methods, that's why I need a mutable map. The mutable map extends the functionality of the map interface by adding that.

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/ui/home/MatchViewModel.kt:
```diff
@@ -24,12 +23,12 @@ class MatchViewModel : ViewModel() {
         _gameService = OfflineGameService(
             gameId,
-            ServiceLocator.getInstance().getFirebaseRepository(),
+            ServiceLocator.getInstance().repository as FirebaseRepository,
```
You cant't cast it that way unless you're surew that it's going to be a FirebaseRepository. You should only do it in the tests or where it's crucially needed.

adam-narozniak replied:
Ok, I've raised the abstraction to Repository

gaetschwartz commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -55,8 +60,7 @@ class OfflineGameService(
         for (pc in computerPlayers) {
-//                uncomment when the proper tests with coroutines implemented
-//            delay(1_000)
+            delay(1_000)
```
This should be customizable, i.e a parameter in OfflineGameService, or something similar.

adam-narozniak replied:
Now, I've pushed the specification of this param to the UI part (where it's default 1_000) so it can be in the future influenced by a user

gaetschwartz commented on app/build.gradle:
```diff
@@ -83,6 +84,7 @@ dependencies {
     testImplementation 'org.hamcrest:hamcrest:2.2'
     implementation 'androidx.navigation:navigation-fragment-ktx:2.4.2'
     implementation 'androidx.navigation:navigation-ui-ktx:2.4.2'
+    testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.6.1'
```
it's already in the dependencies.

adam-narozniak replied:
Now, removed

adam-narozniak commented on app/src/main/java/ch/epfl/sweng/rps/services/OfflineGameService.kt:
```diff
@@ -46,7 +45,13 @@ class OfflineGameService(
+        } else{
+            " "
+        }
+
```
Now, I've changed it to use elvis operator

gaetschwartz commented:
Your OfflineGameService is still not tested enough, only 63%.

adam-narozniak commented:
Now raised it to 89%, left functionality that might be changed later, or I'm not fully sure how they will be used in my case.
